### PR TITLE
Add kubeproxyconfig round trip test

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
This is a follow-up of https://github.com/kubernetes/kubernetes/pull/55961.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
ref: #55959

**Special notes for your reviewer**:
/cc @luxas @mtaufen @ncdc 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+			obj.UDPIdleTimeout = metav1.Duration{Duration: 250 * time.Millisecond}
+			obj.Conntrack.MaxPerCore = utilpointer.Int32Ptr(32 * 1024)
+			obj.Conntrack.Min = utilpointer.Int32Ptr(128 * 1024)
+			obj.IPTables.MasqueradeBit = utilpointer.Int32Ptr(14)
+			obj.Conntrack.TCPEstablishedTimeout = &metav1.Duration{Duration: 24 * time.Hour}
+			obj.Conntrack.TCPCloseWaitTimeout = &metav1.Duration{Duration: 1 * time.Hour}
+			obj.ConfigSyncPeriod.Duration = 15 * time.Minute
+			obj.ClientConnection.ContentType = "application/vnd.kubernetes.protobuf"
+			obj.ClientConnection.QPS = 5.0
+			obj.ClientConnection.Burst = 10
```
Comment (sttts) [inline pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: very static fuzzer :)

Comment (sttts) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: https://xkcd.com/221/

Comment (xiangpengzhao) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: Hmm, these values are from defaulters. I should assign them "random" values here, right? :)

Comment (sttts) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: Fuzzers use random values usually to catch encoding errors.

Comment (xiangpengzhao) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: Thanks for clarifying!

Comment (mtaufen) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: Agreed.
A round trip test with at least defaults is better than nothing.
A round trip test that can assign random valid values within the constraints of each field is even better.
For kubeletconfig we have the former, and need to extend to the latter.

Comment (sttts) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: `c.FuzzNoCustom(obj)` writes random values. Do we have to override them with constant values?

Comment (xiangpengzhao) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: >c.FuzzNoCustom(obj) writes random values. Do we have to override them with constant values?

@mtaufen any thoughts?

**Comment by xiangpengzhao** [comment]
@sttts updated to use random values :)

File: pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+// Funcs returns the fuzzer functions for the kube-proxy apis.
+func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
+	return []interface{}{
+		func(obj *kubeproxyconfig.KubeProxyConfiguration, c fuzz.Continue) {
+			c.FuzzNoCustom(obj)
+			obj.BindAddress = fmt.Sprintf("%d.%d.%d.%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256))
+			obj.HealthzBindAddress = fmt.Sprintf("%d.%d.%d.%d:%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(65536))
+			obj.MetricsBindAddress = fmt.Sprintf("%d.%d.%d.%d:%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(65536))
+			obj.OOMScoreAdj = utilpointer.Int32Ptr(c.Int31())
+			obj.ResourceContainer = "/kube-proxy"
```
Comment (sttts) [inline pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: Aren't most of these randomized already by `c.FuzzNoCustom(obj)`? Usually one only has to fixup some fields which have constraints that `c.FuzzNoCustom(obj)` does not know about, like maybe the addresses above.

Comment (xiangpengzhao) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: @sttts yeah I guess so. But I didn't identify which fields are already randomized. Which type(s) of the fields can't be done by `c.FuzzNoCustom(obj)`?

Comment (sttts) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: > Which type(s) of the fields can't be done by c.FuzzNoCustom(obj)?

Read the source :) Generally it fuzzes quite a lot if not all.

Comment (xiangpengzhao) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: :+1: :smile:

Comment (sttts) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: the two duration below can't be randomized?

Comment (xiangpengzhao) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: yes they can. (last time I tried to randomize them but failed. I believe I missed sth then)

Comment (xiangpengzhao) [reply pkg/proxy/apis/kubeproxyconfig/fuzzer/fuzzer.go]: updated. Local bazel test passed.

**Comment by sttts** [comment]
/assign thocking

For approval.

**Comment by xiangpengzhao** [comment]
typo:)
/assign thockin

**Review by luxas — APPROVED** [review]
/lgtm

Thanks for adding this coverage @xiangpengzhao!

**Comment by xiangpengzhao** [comment]
BUILD files updated. Seems like there was some change of bazel rules.

**Comment by xiangpengzhao** [comment]
/assign @thockin
for approval :)

Merry Xmas all! :christmas_tree:

**Comment by xiangpengzhao** [comment]
Change copyright year to 2018 to fix pull-kubernetes-verify failure. Sorry for bothering folks again.
@sttts @luxas PTAL. ping @thockin for approval.
Thanks folks!

**Pull request merged**